### PR TITLE
Fix non-returnables

### PR DESCRIPTION
**Pull request by pond**
# DRAFT

A variation of #103 after the report of the quite serious bug that on-create/update attributes like `password` were being ignored. See #103 for conversation history - TL;DR, looks like `as_json` was the wrong place and `to_scim` is the right one for the masking out of non-returnable attributes.

### CODE REVIEW

File: app/models/scimitar/resources/base.rb [hunk 1]
```diff
@@ -137,18 +137,21 @@ def constantize_complex_types(hash)
...
 
+      # Renders *in full* as JSON; typically used for write-based operations...
+      #
+      #     record = self.storage_class().new
+      #     record.from_scim!(scim_hash: scim_resource.as_json())
+      #     self.save!(record)
+      #
+      # ...so all fields, even those marked "returned: false", are included.
+      # Use Scimitar::Resources::Mixin::to_scim to obtain a SCIM object with
+      # returnable fields omitted, rendering *that* as JSON via #to_json.
```
Comment (pond) [inline app/models/scimitar/resources/base.rb:148]: This is stil somewhat clear as mud, but at least it gives a fairly strong indication that "you probably don't want to call this for returned JSON data". Between this and the clear examples of controller code in `README.md`, hopefully it's enough for now.

File: app/models/scimitar/resources/base.rb [hunk 2]
```diff
@@ -137,18 +137,21 @@ def constantize_complex_types(hash)
...
 
-        non_returnable_attributes = self.class
-          .schemas
-          .flat_map(&:scim_attributes)
-          .filter_map { |attribute| attribute.name if attribute.returned == 'never' }
-
-        non_returnable_attributes << 'errors'
-
-        original_hash = super(options).except(*non_returnable_attributes)
+        original_hash = super(options).except('errors')
```
Comment (pond) [inline app/models/scimitar/resources/base.rb:154]: Reverts the problematic part of #80.

File: app/models/scimitar/resources/mixin.rb [hunk 3]
```diff
@@ -335,25 +335,16 @@ def scim_queryable_attributes
...
-          attrs_hash      = self.to_scim_backend(data_source: self, attrs_map_or_leaf_value: map)
-          resource        = self.class.scim_resource_type().new(attrs_hash)
-          meta_attrs_hash = { location: location }
-
-          meta_attrs_hash[:created     ] = self.send(timestamps_map[:created     ])&.iso8601(0) if timestamps_map&.key?(:created)
-          meta_attrs_hash[:lastModified] = self.send(timestamps_map[:lastModified])&.iso8601(0) if timestamps_map&.key?(:lastModified)
-
-          resource.meta = Meta.new(meta_attrs_hash)
-          return resource
+          return self.to_scim_for_returning(location: location)
```
Comment (pond) [inline app/models/scimitar/resources/mixin.rb:347]: This gives away that we can't actually fully avoid internally the concept of to-SCIM in "render normally" vs "render everything" mode; more info in later diff comments.

File: app/models/scimitar/resources/mixin.rb [hunk 4]
```diff
@@ -417,7 +408,7 @@ def from_scim!(scim_hash:)
         #
         def from_scim_patch!(patch_hash:)
           frozen_ci_patch_hash = patch_hash.with_indifferent_case_insensitive_access().freeze()
-          ci_scim_hash         = self.to_scim(location: '(unused)').as_json().with_indifferent_case_insensitive_access()
+          ci_scim_hash         = self.to_scim_for_patch_op().as_json().with_indifferent_case_insensitive_access()
```
Comment (pond) [inline app/models/scimitar/resources/mixin.rb:411]: Idiomatic SCIM resource controller code for an `#update` operation has this at the core:

```ruby
record = self.find_record(record_id)
record.from_scim_patch!(patch_hash: patch_hash)
self.save!(record, &block)
# ...and render 'record' in SCIM form...
```

That's how we get to the method being changed here. At the end of this method - you'd need to expand the diff downwards to around new diff line 485 or so - the final steps are:

```ruby
self.from_scim!(scim_hash: ci_scim_hash)
return self
```

The idea is that we build `ci_scim_hash` from the existing data as per the line above, then modify it via whatever update operations are requested, then repopulate the resource from that updated hash. This is related to long-standing issue #6. In lieu of the workload required to produce a true fix, and given that the mixin's tests do include tests to say "if a value is omitted in the Hash then it _is_ set to `nil`", we can't allow `ci_scim_hash` to have missing keys due to non-rendered "returned never" attributes. Instead, it _must_ contain them all, so that the current relatively crude (but still, thanks to SCIM's design, very complex) implementation can work as expected.

I note in passing that in the special case of a password, when something like Devise is in use and being used properly (!) then, once a password is written it is encrypted and cannot be read. Attempts to read the password or confirmation return nil; attempts to _write_ nil or empty string have no side effects and are ignored. This is the _desired outcome_ for that attribute. It does open up the question of non-password "returned never" attributes where the underling model might not allow a read operation, though, causing set-to-nil behaviour in subsequent SCIM patch updates.

TL;DR, what a mess. But this gets us by for now, and resolves issues highlighted by the updated tests.

**Comment by pond** [comment]
@xjunior This is ready to be looked at now. I've annotated the diff with a few comments. It's interesting to see that a reliable, deep fix would only be possible with issue #6 addressed, but as-is, unless people are using very weird schema, it should do just fine for common user and group management workloads.

**Comment by xjunior** [comment]
Awesome, @pond! I have seen the issue #6, but didn't have a chance to deeply analyze it. I plan to do so soon, as we are and will be relying a lot on Scimitar.

File: app/models/scimitar/resources/base.rb (hunk 1 above)
Comment (xjunior) [inline app/models/scimitar/resources/base.rb:148]:
```suggestion
      # non-returnable fields omitted, rendering *that* as JSON via #to_json.
```

**Comment by pond** [comment]
Converted back to draft after a giant rabbit hole opened, in which a hundred yaks in need of shaving.

**Comment by pond** [comment]
@xjunior apart from the suggestion in https://github.com/RIPAGlobal/scimitar/pull/102#issuecomment-2017069630 on #102, which would be useful validation in itself, this is - I think! - ready for review now.

**Review by xjunior — APPROVED** [review]

File: app/models/scimitar/resources/mixin.rb [hunk 5]
```diff
@@ -1108,6 +1307,177 @@ def all_matching_filter(filter:, within_array:, &block)
...
+          # With that done, you can call here with the hash data to be changed
+          # and fragment of attribute map that #find_matching_static_attr_map
+          # (or something like it) found.
+          #
+          # +altering_hash+:: The fragment of SCIM data that might be updated
+          #                   with +nil+ to ultimately lead to an atttribute
+          #                   writer identified through +with_attr_map+ being
+          #                   called with that value. This is often the same
+          #                   that was passed in the +data+ attribute in a
+          #                   in a prior #find_matching_static_attr_map call.
```
Comment (sierra-alpha) [inline app/models/scimitar/resources/mixin.rb:1455]: Classic:

```suggestion
          #                   that was passed in the +data+ attribute in a
          #                   prior #find_matching_static_attr_map call.
```

File: lib/scimitar/support/hash_with_indifferent_case_insensitive_access.rb [hunk 6]
```diff
@@ -49,34 +49,164 @@ module Support
...
+    # * Result:   <tt>'Foo' => true</tt>
+    #
+    # If you're intending to merge nested Hashes, then use ActiveSupport's
+    # #deep_merge or an equivalent. This will have the expected outcome, where
+    # the hash with 'BAR' is _merged_ into the existing value and, therefore,
+    # the original 'Bar' key case is preserved:
+    #
+    # * Original:   <tt>'Foo' => { 'Bar' => 42 }</tt>
+    # * Deep merge: <tt>'FOO' => { 'BAR' => 24 }</tt>
+    # * Result:     <tt>'Foo' => { 'Baf' => 24 }</tt>
```
Comment (sierra-alpha) [inline lib/scimitar/support/hash_with_indifferent_case_insensitive_access.rb:90]: Unless I'm missing something:

```suggestion
    # * Result:     <tt>'Foo' => { 'Bar' => 24 }</tt>
```

**Review by sierra-alpha — COMMENTED** [review]
That casing of the keys work seems like a good head scratch but looks like it's well tested! Just a few typos that I could see.

**Review by sierra-alpha — APPROVED** [review]
